### PR TITLE
USHIFT-1202: Use hostnames over IPs in apiserver when connecting to kubelet

### PR DESCRIPTION
**Pull request by pacevedom**
**Which issue(s) this PR addresses**:

Closes #<Issue Number>

### CODE REVIEW

**Comment by zshi-redhat** [comment]
This was tested to work in my local multi-node env.  LGTM

File: pkg/cmd/init.go [hunk 1]
```diff
@@ -181,7 +181,12 @@ func certSetup(cfg *config.Config) (*certchains.CertificateChains, error) {
 						Name:         "kubelet-server",
 						ValidityDays: cryptomaterial.ShortLivedCertificateValidityDays,
 					},
-					Hostnames: []string{cfg.Node.HostnameOverride, cfg.Node.NodeIP},
+					// Kubelet hostnames must contain those addresses advertised by a node. In MicroShift these
+					// correspond to the node's IP address and the hostname. In order to accommodate having more
+					// than one node for CNCF certification tests, the subject alternative names from the
```
Comment (dhellmann) [inline pkg/cmd/init.go:186]: Let's not mention the tests here. We may have other reasons to deal with multi-node in the future.

```suggestion
					// than one node, the subject alternative names from the
```

File: pkg/controllers/kube-apiserver.go [hunk 2]
```diff
@@ -125,7 +125,7 @@ func (s *KubeAPIServer) configure(cfg *config.Config) error {
 			"kubelet-certificate-authority":   {cryptomaterial.CABundlePath(kubeCSRSignerDir)},
 			"kubelet-client-certificate":      {cryptomaterial.ClientCertPath(kubeletClientDir)},
 			"kubelet-client-key":              {cryptomaterial.ClientKeyPath(kubeletClientDir)},
-			"kubelet-preferred-address-types": {"InternalIP", "Hostname"},
+			"kubelet-preferred-address-types": {"Hostname", "InternalIP"},
```
Comment (dhellmann) [inline pkg/controllers/kube-apiserver.go:128]: A comment here (or in the commit message) explaining how this change fixes the problem would help future maintainers.

**Review by dhellmann — COMMENTED** [review]
The commit messages still say WIP.

File: pkg/cmd/init.go (hunk 1 above)
Comment (pacevedom) [reply pkg/cmd/init.go]: Done

File: pkg/controllers/kube-apiserver.go (hunk 2 above)
Comment (pacevedom) [reply pkg/controllers/kube-apiserver.go:135]: Done

**Review by dhellmann — COMMENTED** [review]
The implementation and new comments look OK, but it wasn't clear from the slack thread whether we still need this. Zenghui and Jeremy should take a look.

/assign @zshi-redhat @jerpeter1

File: pkg/controllers/kube-apiserver.go [hunk 3]
```diff
@@ -125,7 +125,14 @@ func (s *KubeAPIServer) configure(cfg *config.Config) error {
 			"kubelet-certificate-authority":   {cryptomaterial.CABundlePath(kubeCSRSignerDir)},
 			"kubelet-client-certificate":      {cryptomaterial.ClientCertPath(kubeletClientDir)},
 			"kubelet-client-key":              {cryptomaterial.ClientKeyPath(kubeletClientDir)},
-			"kubelet-preferred-address-types": {"InternalIP", "Hostname"},
+			// MicroShift nodes expose these two types of addresses. In order to support having more than one
+			// node with the current approach (which is running a stand alone kubelet and share certificates
+			// with the master node) we need to use names only because of the way certificates are generated.
+			// The names are fed through the subjectAltNames configuration parameter, and using master node
```
[the remark this answers is not in the document]
Comment (pacevedom) [reply pkg/controllers/kube-apiserver.go:131]: Yes, it is part of the current approach. If we are to generate the certificates and then manually signing them with the CA (for which you need root access to microshift) then it should not be needed. But that is a bit more complex.

File: pkg/controllers/kube-apiserver.go [hunk 4]
```diff
@@ -125,7 +125,14 @@ func (s *KubeAPIServer) configure(cfg *config.Config) error {
 			"kubelet-certificate-authority":   {cryptomaterial.CABundlePath(kubeCSRSignerDir)},
 			"kubelet-client-certificate":      {cryptomaterial.ClientCertPath(kubeletClientDir)},
 			"kubelet-client-key":              {cryptomaterial.ClientKeyPath(kubeletClientDir)},
-			"kubelet-preferred-address-types": {"InternalIP", "Hostname"},
+			// MicroShift nodes expose these two types of addresses. In order to support having more than one
```
Comment (jerpeter1) [inline pkg/controllers/kube-apiserver.go:128]: The current approach (using standalone kubelet) does not share certs with the master node, it uses a cert with cluster-admin privileges to bootstrap: https://kubernetes.io/docs/reference/access-authn-authz/kubelet-tls-bootstrapping/

**Comment by zshi-redhat** [comment]
> The implementation and new comments look OK, but it wasn't clear from the slack thread whether we still need this. Zenghui and Jeremy should take a look.
> 
> /assign @zshi-redhat @jerpeter1

I think we still want the first commit, which is a bug fix according to @pacevedom, not specific to multi-node.
Whether second commit is required depending on how kubelet runs on the extra worker node, will have a follow-up discussion.

File: pkg/controllers/kube-apiserver.go (hunk 4 above)
Comment (pacevedom) [reply pkg/controllers/kube-apiserver.go:128]: In light of this I will remove the second commit, as its handled correctly by bootstrapping.

**Comment by jerpeter1** [comment]
I had to make a minor adjustment to my script, but I tested this and it works as intended.

/lgtm

**Comment by pacevedom** [comment]
/override ci/prow/e2e-openshift-conformance-reduced-arm
/override ci/prow/microshift-e2e-arm
arm jobs are flaky due to a kernel bug.
/override ci/prow/e2e-openshift-conformance-reduced
Overriding conformance reduced as the failure relates to 1.27 rebase. Should this PR fail then any test trying to retrieve anything from the kubelet would fail, which is not the case.

**Pull request merged**